### PR TITLE
Split up search.cu to improve compile time

### PR DESCRIPTION
**Pull request by davidwendt**
Splits up `src/search/search.cu` to improve its compile time. The `cudf::upper_bound` and `cudf::lower_bound` API implementations are moved to `search_ordered.cu` leaving the `cudf::contains` API implementations in `search.cu`. Further the `struct` column type specialization implementation is moved to a new `structs/search/search.cu` source file. No other refactoring was required.

Current compile time in CI is about 15 minutes:
https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-cpu-cuda-build/CUDA=11.5/9335/Build_20Metrics_20Report/

New compile time is about 11 minutes: (27% improvement)
https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-cpu-cuda-build/CUDA=11.5/9343/Build_20Metrics_20Report/

Hopefully this pattern can be used when adding additional search capabilities (e.g. #10656) as well.

### CODE REVIEW

**Review by vyasr — APPROVED** [review]

File: cpp/include/cudf/structs/detail/search.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * @brief Returns true if the scalar is found in the column.
+ *
+ * @param haystack Column to search against
+ * @param needle Scalar to search for
+ * @param stream  CUDA stream used for device memory operations and kernel launches.
+ * @return True if the `needle` is found in `haystack`
+ */
+bool contains(structs_column_view const& haystack,
```
Comment (harrism) [inline cpp/include/cudf/structs/detail/search.hpp:35]: If this is the only thing here, why not call this contains.hpp?

File: cpp/src/search/search.cu [hunk 2]
```diff
@@ -351,52 +220,8 @@ std::unique_ptr<column> contains(column_view const& haystack,
...
-                                    std::vector<order> const& column_order,
-                                    std::vector<null_order> const& null_precedence,
-                                    rmm::mr::device_memory_resource* mr)
-{
-  CUDF_FUNC_RANGE();
-  return detail::upper_bound(
-    haystack, needles, column_order, null_precedence, rmm::cuda_stream_default, mr);
-}
-
 bool contains(column_view const& haystack, scalar const& needle)
```
Comment (harrism) [inline cpp/src/search/search.cu:225]: Maybe rename this to contains.cu?

File: cpp/src/structs/search/search.cu [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/find.h>
+#include <thrust/iterator/counting_iterator.h>
+
+namespace cudf {
+namespace structs {
+namespace detail {
+
+bool contains(structs_column_view const& haystack,
```
Comment (harrism) [inline cpp/src/structs/search/search.cu:34]: Maybe rename file to contains.cu?

**Review by harrism — COMMENTED** [review]
Looks good. Just some file renaming suggestions.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/search/search_ordered.cu [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cudf {
+namespace detail {
+
+std::unique_ptr<column> search_ordered(table_view const& haystack,
```
Comment (bdice) [inline cpp/src/search/search_ordered.cu:35]: This function could be in an anonymous namespace, if desired.

**Review by bdice — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Review by Ethyling — APPROVED** [review]

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**